src/utils/components/withMatchMedia.jsx

```diff
@@ -0,0 +1,111 @@
+import React from 'react';
+import { MEDIA_QUERIES } from '../designConstants';
+
+const breakpointNames = Object.keys(MEDIA_QUERIES);
+
+/**
+ * @param {String} breakpoint - name of breakpoint (for example, 'medium')
+ * @returns {String} - state property name (for example, isAtMediumUp)
+ */
+export const getStateNameByBreakpoint = breakpoint => {
+	const capitalizedBp = `${breakpoint.substr(0,1).toUpperCase()}${breakpoint.substr(1)}`;
+	return `isAt${capitalizedBp}Up`;
+};
+
+/**
+ * @param {Array} mediaQueries - list of matchMedia-created MediaQueryList objects
+ * @param {Array} breakpoints - array of breakpoint names that were passed to HOC
+ * @returns {Object} - updated `state` object for `withMatchMedia`
+ */
+export const getUpdatedMediaState = (mediaQueries) => mediaQueries
+	.reduce((state, mq, i) => {
+		state[getStateNameByBreakpoint(breakpointNames[i])] = mq.matches;
+		return state;
+	}, {});
+
+
+/**
+ * Provides viewport-aware props to wrapped component.
+ *
+ * @param {React.element} WrappedComponent - the component to wrap
+ * @param {Array} breakpoints - array of breakpoint names to watch
+ */
+export const withMatchMedia = (
+	WrappedComponent
+) => class extends React.Component {
+	/**
+	 * @constructor
+	 * @param {Object} - React element props
+	 */
+	constructor(props) {
+		super(props);
+
+		// map breakpoint names to MediaQueryList objects...
+		if (window && window.matchMedia) {
```

Comment:
This state assignment really needs to go into `componentDidMount` even though it will trigger a re-render. The unavoidable problem is that React expects the initially-generated browser markup to exactly match the server-generated markup, and if the `media` prop is used to affect markup in the browser but not the server, then there will be (potential) problems. Apparently they are going to change this behavior in the next major version of React, but I think we can pay the (small) cost of a re-render for now and put it in the 'correct' place in `componentDidMount`